File: sandctl-ts/src/commands/resize.ts
import { confirm } from "@inquirer/prompts";
import { Command } from "commander";

import {
	type Config,
	getProviderConfig,
	load,
	type ProviderConfig,
} from "@/config/config";
import type { HetznerProvider } from "@/hetzner/provider";
import { get as getProviderFromRegistry } from "@/provider/registry";
import { normalizeName, validateID } from "@/session/id";
import { SessionStore } from "@/session/store";
import { NotFoundError } from "@/session/types";

import { CommandExitError } from "@/commands/destroy";

interface Dependencies {
	loadConfig: (configPath?: string) => Promise<Config>;
	resolveProvider: (
		name: string,
		config: ProviderConfig,
	) => ReturnType<typeof getProviderFromRegistry>;
}

const defaultDependencies: Dependencies = {
	loadConfig: load,
	resolveProvider: getProviderFromRegistry,
};

export async function runResize(
	name: string,
	serverType: string,
	options: { force: boolean; upgradeDisk: boolean },
	store = new SessionStore(),
	deps: Partial<Dependencies> = {},
	configPath?: string,
): Promise<void> {
	const dependencies = {
		...defaultDependencies,
		...deps,
	};

	const normalized = normalizeName(name);
	if (!validateID(normalized)) {
		throw new Error(`invalid session name format: ${name}`);
	}

	const session = await store.get(normalized).catch((error: unknown) => {
		if (error instanceof NotFoundError) {
			throw new CommandExitError(
				`Session '${normalized}' not found. Use 'sandctl list' to see available sessions.`,
				4,
			);
		}
		throw error;
	});

	if (!session.provider_id) {
		throw new Error(
			`Session '${session.id}' is in legacy format and cannot be resized.`,
		);
	}

	const config = await dependencies.loadConfig(configPath);
	const providerConfig = getProviderConfig(config, session.provider);
	if (!providerConfig) {
		throw new Error(`Provider '${session.provider}' is not configured.`);
	}

	const provider = dependencies.resolveProvider(
		session.provider,
		providerConfig,
	);

	// Check provider supports resize
	if (!("resize" in provider) || typeof provider.resize !== "function") {
		throw new Error(
			`Provider '${session.provider}' does not support resize.`,
		);
	}

	// Confirm unless --force
	if (!options.force) {
		let message = `Resize session '${session.id}' to ${serverType}?`;
		if (options.upgradeDisk) {
			console.warn(
				`Warning: --upgrade-disk will expand the disk to match ${serverType}. This prevents future downgrades.`,
			);
			message = `Resize session '${session.id}' to ${serverType} (with disk upgrade)?`;
		}

		const accepted = await confirm({ message, default: false });
		if (!accepted) {
			console.log("Canceled.");
			return;
		}
	}

	console.log(
		`Resizing session '${session.id}' to ${serverType}...`,
	);

	await (provider as HetznerProvider).resize(
		session.provider_id,
		serverType,
		options.upgradeDisk,
	);

	// Update session store with new server type
	await store.update(session.id, { server_type: serverType });

	console.log(
		`Session '${session.id}' resized to ${serverType}.`,
	);
}

export function registerResizeCommand(): Command {
	return new Command("resize")
		.description("Resize a session's server type (CPU/RAM)")
		.argument("<name>", "Session name")
		.argument("<server-type>", "Target server type (e.g., cpx41)")
		.option("-f, --force", "Skip confirmation prompt", false)
		.option(
			"--upgrade-disk",
			"Expand disk to match new server type (prevents downgrades)",
			false,
		)
		.action(
			async (
				name: string,
				serverType: string,
				options: { force: boolean; upgradeDisk: boolean },
				command,
			) => {
				const globals = command.optsWithGlobals() as { config?: string };
				await runResize(
					name,
					serverType,
					options,
					undefined,
					undefined,
					globals.config,
				);
			},
		);
}

File: internal/cli/resize.go
package cli

import (
	"context"
	"errors"
	"fmt"
	"os"

	"github.com/spf13/cobra"

	"github.com/sandctl/sandctl/internal/provider"
	"github.com/sandctl/sandctl/internal/session"
	"github.com/sandctl/sandctl/internal/ui"
)

var (
	resizeForce      bool
	resizeUpgradeDisk bool
)

var resizeCmd = &cobra.Command{
	Use:   "resize <name> <server-type>",
	Short: "Resize a session's server type (CPU/RAM)",
	Long: `Resize a sandboxed VM by changing its Hetzner server type.

The server will be stopped, resized, and restarted automatically.

By default, the disk size is not changed (allowing future downgrades).
Use --upgrade-disk to expand the disk to match the new server type
(prevents future downgrades).`,
	Example: `  # Resize to a larger server type
  sandctl resize alice cpx41

  # Resize with disk upgrade (prevents future downgrades)
  sandctl resize alice cpx41 --upgrade-disk

  # Resize without confirmation
  sandctl resize alice cpx41 --force`,
	Args: cobra.ExactArgs(2),
	RunE: runResize,
}

func init() {
	resizeCmd.Flags().BoolVarP(&resizeForce, "force", "f", false, "skip confirmation prompt")
	resizeCmd.Flags().BoolVar(&resizeUpgradeDisk, "upgrade-disk", false, "expand disk to match new server type (prevents downgrades)")

	rootCmd.AddCommand(resizeCmd)
}

func runResize(cmd *cobra.Command, args []string) error {
	ctx := context.Background()

	sessionName := session.NormalizeName(args[0])
	serverType := args[1]

	// Validate session name format
	if !session.ValidateID(sessionName) {
		return fmt.Errorf("invalid session name format: %s", args[0])
	}

	// Get session from store
	store := getSessionStore()
	sess, err := store.Get(sessionName)
	if err != nil {
		var notFound *session.NotFoundError
		if errors.As(err, &notFound) {
			ui.PrintError(os.Stderr, "session '%s' not found", sessionName)
			fmt.Fprintln(os.Stderr)
			fmt.Fprintln(os.Stderr, "Use 'sandctl list' to see active sessions.")
			return nil
		}
		return err
	}

	// Check for legacy sessions
	if sess.IsLegacySession() {
		ui.PrintError(os.Stderr, "session '%s' is from an old version and cannot be resized", sessionName)
		return nil
	}

	// Get provider
	prov, err := getProviderFromSession(sess)
	if err != nil {
		return fmt.Errorf("failed to get provider: %w", err)
	}

	// Check provider supports resize
	resizer, ok := prov.(provider.Resizer)
	if !ok {
		return fmt.Errorf("provider '%s' does not support resize", sess.Provider)
	}

	// Confirm unless --force
	if !resizeForce {
		msg := fmt.Sprintf("Resize session '%s' to %s?", sessionName, serverType)
		if resizeUpgradeDisk {
			ui.PrintWarning(os.Stdout, "--upgrade-disk will expand the disk to match %s. This prevents future downgrades.", serverType)
			msg = fmt.Sprintf("Resize session '%s' to %s (with disk upgrade)?", sessionName, serverType)
		}
		confirmed, confirmErr := ui.Confirm(os.Stdin, os.Stdout, msg)
		if confirmErr != nil {
			return fmt.Errorf("failed to read confirmation: %w", confirmErr)
		}
		if !confirmed {
			fmt.Println("Canceled.")
			return nil
		}
	}

	// Run resize with progress steps
	err = ui.RunSteps(os.Stdout, []ui.ProgressStep{
		{
			Message: "Resizing session",
			Action: func() error {
				return resizer.Resize(ctx, sess.ProviderID, serverType, resizeUpgradeDisk)
			},
		},
	})
	if err != nil {
		return fmt.Errorf("resize failed: %w", err)
	}

	fmt.Printf("Session '%s' resized to %s.\n", sessionName, serverType)

	return nil
}
